### PR TITLE
Add support for remaining Credentials

### PR DESCRIPTION
**Pull request by cmelchior**
Except Custom functions (which require BSON support) and Google Auth Code (which require support being added to the C-API).

### CODE REVIEW

**Comment by clementetb** [comment]
LGTM 🚀

**Review by clementetb — APPROVED** [review]
LGTM 🚀

File: packages/cinterop/src/darwin/kotlin/io/realm/internal/interop/RealmInterop.kt [hunk 1]
```diff
@@ -1540,6 +1540,41 @@ actual object RealmInterop {
...
+    }
+
+    actual fun realm_app_credentials_new_google_id_token(idToken: String): NativePointer {
+        memScoped {
+            return CPointerWrapper(realm_wrapper.realm_app_credentials_new_google(idToken))
+        }
+    }
+
+    actual fun realm_app_credentials_new_google_auth_code(authCode: String): NativePointer {
+        TODO("See ttps://github.com/realm/realm-core/issues/5347")
```
Comment (nhachicha) [inline packages/cinterop/src/darwin/kotlin/io/realm/internal/interop/RealmInterop.kt:1568]:
```suggestion
        TODO("See https://github.com/realm/realm-core/issues/5347")
```

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/Credentials.kt [hunk 2]
```diff
@@ -52,5 +66,68 @@ public interface Credentials {
         public fun emailPassword(email: String, password: String): Credentials {
             return CredentialImpl(CredentialImpl.emailPassword(email, password))
         }
+
+        /**
+         * Creates credentials representing a login using a user API key.
```
Comment (nhachicha) [inline packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/Credentials.kt:71]: should we link/explain who provides/creates the given `key`?

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/CredentialsTests.kt [hunk 3]
```diff
@@ -45,68 +49,82 @@ class CredentialsTests {
...
-                AuthenticationProvider.EMAIL_PASSWORD ->
-                    emailPassword()
-                else -> error("Untested credentials type: $value")
+            val credentials: Credentials = when (value) {
+                AuthenticationProvider.ANONYMOUS -> anonymous()
+                AuthenticationProvider.EMAIL_PASSWORD -> emailPassword()
+                AuthenticationProvider.API_KEY -> apiKey()
+                AuthenticationProvider.APPLE -> apple()
+                AuthenticationProvider.FACEBOOK -> facebook()
+                AuthenticationProvider.GOOGLE -> { continue /* Ignore, see below */ }
```
Comment (nhachicha) [inline test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/CredentialsTests.kt:58]: shouldn't we disable Google Auth until the core issue is implemented?

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/CredentialsTests.kt [hunk 4]
```diff
@@ -45,68 +49,82 @@ class CredentialsTests {
...
-//        assertTrue(creds.asJson().contains(id.toString()))
-//    }
-//
-//    @Test
-//    fun customFunction_invalidInput() {
-//        assertFailsWith<IllegalArgumentException> { Credentials.customFunction(null) }
-//    }
+    private fun apiKey(): Credentials {
+        val creds: Credentials = Credentials.apiKey("token")
+        // assertTrue(creds.asJson().contains("token")) // Treat the JSON as an opaque value.
```
Comment (nhachicha) [inline test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/CredentialsTests.kt:103]: why commented?

**Review by nhachicha — APPROVED** [review]
LGTM not sure though why a lot of test are still commented out (besides the Google provider one) https://github.com/realm/realm-kotlin/pull/743/files#diff-60ddcacd3669ce7ec2d1a6b9d0c0de18be6608ed66007ed368c540793c47437eR166

File: packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/Credentials.kt (hunk 2 above)
Comment (cmelchior) [reply packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/Credentials.kt:71]: I have added a link to the docs.

File: test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/CredentialsTests.kt (hunk 4 above)
Comment (cmelchior) [reply test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/CredentialsTests.kt:104]: It is waiting for https://github.com/realm/realm-core/issues/5348, I have added a comment in the code about it

**Pull request merged**